radix-engine/src/engine/substate_receipt.rs

```diff
@@ -2,60 +2,91 @@ use sbor::*;
+use crate::engine::SubstateParentId;
+use crate::engine::track::{VirtualSubstateId};
 
 use crate::ledger::*;
 
 
 pub struct CommitReceipt {
-    pub down_substates: HashSet<(Hash, u32)>,
-    pub up_substates: Vec<(Hash, u32)>,
+    pub virtual_down_substates: HashSet<HardVirtualSubstateId>,
+    pub down_substates: HashSet<PhysicalSubstateId>,
+    pub virtual_up_substates: Vec<PhysicalSubstateId>,
+    pub up_substates: Vec<PhysicalSubstateId>,
 }
 
 impl CommitReceipt {
     fn new() -> Self {
         CommitReceipt {
+            virtual_down_substates: HashSet::new(),
             down_substates: HashSet::new(),
+            virtual_up_substates: Vec::new(),
             up_substates: Vec::new(),
         }
     }
 
-    fn down(&mut self, id: (Hash, u32)) {
+    fn virtual_down(&mut self, id: HardVirtualSubstateId) {
+        self.virtual_down_substates.insert(id);
+    }
+
+    fn down(&mut self, id: PhysicalSubstateId) {
         self.down_substates.insert(id);
     }
 
-    fn up(&mut self, id: (Hash, u32)) {
+    fn virtual_space_up(&mut self, id: PhysicalSubstateId) {
+        self.up_substates.push(id);
+    }
+
+    fn up(&mut self, id: PhysicalSubstateId) {
         self.up_substates.push(id);
     }
 }
 
+#[derive(Debug, Clone, Hash, TypeId, Encode, Decode, PartialEq, Eq)]
+pub struct HardVirtualSubstateId(PhysicalSubstateId, Vec<u8>);
+
 #[derive(Debug, Clone, TypeId, Encode, Decode, PartialEq, Eq)]
-pub enum StateUpdateInstruction {
-    Down(Hash, u32),
+pub enum SubstateOperation {
+    VirtualDown(VirtualSubstateId),
+    Down(PhysicalSubstateId),
+    VirtualUp(Vec<u8>),
     Up(Vec<u8>, Vec<u8>),
 }
 
 #[derive(Debug, Clone, TypeId, Encode, Decode, PartialEq, Eq)]
-pub struct StateUpdateReceipt {
-    pub instructions: Vec<StateUpdateInstruction>,
+pub struct SubstateOperationsReceipt {
+    pub substate_operations: Vec<SubstateOperation>,
 }
 
-impl StateUpdateReceipt {
+impl SubstateOperationsReceipt {
     /// Commits changes to the underlying ledger.
     /// Currently none of these objects are deleted so all commits are puts
     pub fn commit<S: WriteableSubstateStore>(mut self, store: &mut S) -> CommitReceipt {
         let hash = hash(scrypto_encode(&self));
         let mut receipt = CommitReceipt::new();
         let mut id_gen = SubstateIdGenerator::new(hash);
 
-        for instruction in self.instructions.drain(RangeFull) {
+        for instruction in self.substate_operations.drain(RangeFull) {
             match instruction {
-                StateUpdateInstruction::Down(hash, index) => receipt.down((hash, index)),
-                StateUpdateInstruction::Up(key, value) => {
+                SubstateOperation::VirtualDown(VirtualSubstateId(parent_id, key)) => {
+                    let parent_hard_id = match parent_id {
+                        SubstateParentId::Exists(real_id) => real_id,
+                        SubstateParentId::New(index) => PhysicalSubstateId(hash, index.try_into().unwrap()),
```

Comment:
Just so I understand this correctly, the `SubstateParentId::New` condition is to enable using virtual parent created in the same transaction?